oak-run-commons/src/main/java/org/apache/jackrabbit/oak/index/indexer/document/incrementalstore/IncrementalFlatFileStoreStrategy.java

```diff
@@ -0,0 +1,181 @@
+    private final boolean deleteOriginal = Boolean.parseBoolean(System.getProperty(OAK_INDEXER_DELETE_ORIGINAL, "true"));
+    private final int maxMemory = Integer.getInteger(OAK_INDEXER_MAX_SORT_MEMORY_IN_GB, OAK_INDEXER_MAX_SORT_MEMORY_IN_GB_DEFAULT);
+    private long textSize = 0;
+    private long entryCount = 0;
+    private final Set<String> preferredPathElements;
+
+    public IncrementalFlatFileStoreStrategy(NodeStore nodeStore, @NotNull String beforeCheckpoint, @NotNull String afterCheckpoint, File storeDir,
+                                            Set<String> preferredPathElements, Compression algorithm, Predicate<String> pathPredicate, NodeStateEntryWriter entryWriter) {
+        this.nodeStore = nodeStore;
+        this.beforeCheckpoint = beforeCheckpoint;
+        this.afterCheckpoint = afterCheckpoint;
+        this.storeDir = storeDir;
+        this.algorithm = algorithm;
+        this.pathPredicate = pathPredicate;
+        this.entryWriter = entryWriter;
+        this.preferredPathElements = preferredPathElements;
+        this.comparator = new PathElementComparator(preferredPathElements);
+    }
+
+    @Override
+    public File createSortedStoreFile() throws IOException {
+        Stopwatch sw = Stopwatch.createStarted();
+        File file = new File(storeDir, getSortedStoreFileName(algorithm));
+        try (BufferedWriter w = FlatFileStoreUtils.createWriter(file, algorithm)) {
+            NodeState before = Objects.requireNonNull(nodeStore.retrieve(beforeCheckpoint));
+            NodeState after = Objects.requireNonNull(nodeStore.retrieve(afterCheckpoint));
+            EditorDiff.process(VisibleEditor.wrap(new IncrementalFlatFileStoreEditor(w, entryWriter, pathPredicate, this)), before, after);
+        }
+        String sizeStr = algorithm == null || algorithm.equals(Compression.NONE) ? "" : String.format("compressed/%s actual size", humanReadableByteCount(textSize));
+        log.info("Dumped {} nodestates in json format in {} ({} {})", entryCount, sw, humanReadableByteCount(file.length()), sizeStr);
+        return sortStoreFile(file);
+    }
+
+    @Override
+    public File createMetadataFile() throws IOException {
+        File metadataFile = new File(storeDir, getMetadataFileName(algorithm));
+        IncrementalIndexStoreMetadata indexStoreMetadata = new IncrementalIndexStoreMetadata(this);
+        try (BufferedWriter metadataWriter = FlatFileStoreUtils.createWriter(metadataFile, algorithm)) {
+            writeMetadataToFile(metadataWriter, indexStoreMetadata);
+        }
+        log.info("Created metadataFile:{} with strategy:{} ", metadataFile.getPath(), indexStoreMetadata.getStoreType());
+        return metadataFile;
+    }
+
+
+    private void writeMetadataToFile(BufferedWriter w, IncrementalIndexStoreMetadata indexStoreMetadata) throws IOException {
+        ObjectMapper mapper = new ObjectMapper();
+        mapper.writeValue(w, indexStoreMetadata);
+    }
```

Comment:
Lets discuss this.

Reply:
Added generics to avoid code duplication. Though the generics type is open and there is no condition on what datatype it should be. Though we can add another interface if this doesn't feel right.